accord-core/src/main/java/accord/local/Commands.java

```diff
@@ -900,47 +978,98 @@ else if (cur.has(until))
                     break;
                 }
 
-                TxnId directlyBlockedOnCommit = firstWaitingOnCommit(cur);
-                TxnId directlyBlockedOnApply = firstWaitingOnApply(cur, directlyBlockedOnCommit);
+                WaitingOn waitingOn = cur.hasBeen(Committed) ? cur.asCommitted().waitingOn() : WaitingOn.EMPTY;
+                TxnId directlyBlockedOnCommit;
+                TxnId directlyBlockedOnApply = waitingOn.nextWaitingOnApply();
                 if (directlyBlockedOnApply != null)
                 {
+                    // preferentially block on apply, as this probably saves us additional bookkeeping
+                    // by giving other dependencies time to complete
                     push(directlyBlockedOnApply, Known.Done);
+                    prevSafe = curSafe;
                 }
-                else if (directlyBlockedOnCommit != null)
+                else if ((directlyBlockedOnCommit = waitingOn.nextWaitingOnCommit()) != null)
                 {
                     push(directlyBlockedOnCommit, ExecuteAtOnly);
+                    prevSafe = curSafe;
                 }
                 else
                 {
-                    if (cur.hasBeen(Committed) && !cur.hasBeen(ReadyToExecute) && !cur.asCommitted().isWaitingOnDependency())
+                    if (cur.hasBeen(Committed) && !cur.is(ReadyToExecute) && !cur.is(Applying) && !cur.asCommitted().isWaitingOnDependency())
                     {
-                        if (!maybeExecute(safeStore, curSafe, progressShard(safeStore, cur), false, false))
+                        if (!maybeExecute(safeStore, curSafe, cur.progressShard(), false, false))
                             throw new AssertionError("Is able to Apply, but has not done so");
                         // loop and re-test the command's status; we may still want to notify blocking, esp. if not homeShard
                         continue;
                     }
 
-                    Unseekables<?, ?> someKeys = cur.maxUnseekables();
-                    if (someKeys == null && prev != null) someKeys = prev.partialDeps().someUnseekables(cur.txnId());
-                    Invariants.checkState(someKeys != null);
-                    logger.trace("{} blocked on {} until {}", txnIds[0], cur.txnId(), until);
-                    safeStore.progressLog().waiting(cur.txnId(), until, someKeys);
-                    return;
+                    Participants<?> someParticipants = null;
+                    if (cur.route() != null)
+                    {
+                        someParticipants = cur.route().participants();
+                    }
+                    else if (prev != null)
+                    {
+                        someParticipants = prev.partialDeps().participants(cur.txnId());
```

Comment:
can't `prev.partialDeps().participants(cur.txnId());` NPE?  if we depend on something that was truncated (but not yet purged), or `NotDefined`?

Reply:
I think you may have confused about the meaning of `prev` here - we're exploring dependencies, so `prev` is the one that depends on `cur`, and we check for this at the start, stepping up the stack as we find things have been applied/truncated.

Reply:
ah, I had this backwards... thought `prev` was this txn dependency and not the other way around.